test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetSourcesTests.cs

```diff
@@ -453,6 +461,372 @@ public void TestCommandInvalidArguments(string command, int badCommandIndex)
+
+                // Act
+                CommandRunnerResult result = _fixture.RunDotnet(projectDirectory, $"add {projectFilePath} package {packageX} -s {sourceRelativePath}", ignoreExitCode: true);
+
+                // Assert
+                // It should fail because there is no packageX, only packageY in source directory.
+                result.Success.Should().BeFalse(because: result.AllOutput);
+            }
+        }
+
+        [Theory]
+        [InlineData("net5.0")]
+        public async Task AddPkg_WithV2_RelativePath_VersionSpecified_Success(string targetFrameworks)
+        {
+            using (var pathContext = new SimpleTestPathContext())
+            {
+                var projectName = "projectA";
+                SimpleTestProjectContext projectA = XPlatTestUtils.CreateProject(projectName, pathContext, targetFrameworks);
+                var packageX = "packageX";
+                var packageX_V1 = new PackageIdentity(packageX, new NuGetVersion("1.0.0"));
+                var packageX_V2 = new PackageIdentity(packageX, new NuGetVersion("2.0.0"));
+                var packageXPath = Path.Combine(pathContext.WorkingDirectory, "Custompackages");
+                var sourceRelativePath = RuntimeEnvironmentHelper.IsWindows ? "..\\..\\Custompackages" : "../../Custompackages";
+
+                // Generate Package
+                await SimpleTestPackageUtility.CreateFolderFeedV2Async(
+                    packageXPath, //not using solution source folder
+                    new PackageIdentity[] { packageX_V1, packageX_V2 });
+
+                var projectDirectory = Path.Combine(pathContext.SolutionRoot, projectName);
+                var projectFilePath = Path.Combine(projectDirectory, $"{projectName}.csproj");
+
+                // Act
+                CommandRunnerResult result = _fixture.RunDotnet(projectDirectory, $"add {projectFilePath} package {packageX} -s {sourceRelativePath} -v {packageX_V1.Version}", ignoreExitCode: true);
+
+                // Assert
+                result.Success.Should().BeTrue(because: result.AllOutput);
+
+                // Make sure source is replaced in generated dgSpec file.
+                PackageSpec packageSpec = projectA.AssetsFile.PackageSpec;
+                string[] sources = packageSpec.RestoreMetadata.Sources.Select(s => s.Name).ToArray();
+                Assert.Equal(sources.Count(), 1);
+                Assert.True(sources[0].Contains("Custompackages"));
+
+                var ridlessTarget = projectA.AssetsFile.Targets.Where(e => string.IsNullOrEmpty(e.RuntimeIdentifier)).Single();
+                ridlessTarget.Libraries.Should().Contain(e => e.Type == "package" && e.Name == packageX);
+                // Should resolve to specified version.
+                ridlessTarget.Libraries.Should().Contain(e => e.Version.Equals(packageX_V1.Version));
+            }
+        }
+
+        [Theory]
```

Comment:
If you are going to write a theory, you can probably combine the test cases. 

For example, the theory data can be the version.
```cs
[InlineDate("")]
[InlineDate("1.0.0")]
```

Reply:
I changed code to 'Fact'. Since dotnet integration test are slow I don't want overburden with many input cases.

Reply:
> I changed code to 'Fact'. Since dotnet integration test are slow I don't want overburden with many input cases.


I was suggesting you merge test cases into 1. I think having a test for scenarios with and without version makes sense.